src/next.rs

```diff
@@ -0,0 +1,35 @@
+//! Next futures for `Body`
+
+use crate::Body;
+
+use core::future::Future;
+use core::pin::Pin;
+use core::task;
+
+#[derive(Debug)]
+///Future that resolves to the next data chunk from `Body`
+pub struct NextData<'a, T>(pub(crate) &'a mut T);
+
+impl<'a, T: Body + Unpin> Future for NextData<'a, T> {
+    type Output = Option<Result<T::Data, T::Error>>;
+
+    fn poll(self: Pin<&mut Self>, ctx: &mut task::Context<'_>) -> task::Poll<Self::Output> {
+        let body = unsafe { self.map_unchecked_mut(|this| &mut this.0) };
+
+        Body::poll_data(body, ctx)
+    }
+}
+
+#[derive(Debug)]
+///Future that resolves to the optional trailers from `Body`
+pub struct NextTrailers<'a, T>(pub(crate) &'a mut T);
+
+impl<'a, T: Body + Unpin> Future for NextTrailers<'a, T> {
```

Comment:
why does this T need Unpin?

Reply:
because reference?

Reply:
probably can be removed, but on other hand it is not per se necessary

Reply:
I think we should bring in pin-project and use that here so that we can accept an Unpin T.

Reply:
Im not sure what you mean but I see no need for pin-project

Reply:
`Unpin` comes from this default impl https://github.com/hyperium/http-body/blob/master/src/lib.rs#L70

So I think it is fine

Reply:
That is specifically for a Unpin type, this is how std future is done. It allows you to finally poll the future once you've gotten it through a Pin via Pin unchecked etc. So we shouldn't add the bound if we don't need it imo.

Reply:
the bound is needed, because `Next` use reference

```
  --> src\next.rs:19:9
   |
19 |         Body::poll_data(body, ctx)
   |         ^^^^^^^^^^^^^^^ the trait `std::marker::Unpin` is not implemented for `T`
   |
   = help: consider adding a `where T: std::marker::Unpin` bound
   = note: required because of the requirements on the impl of `Body` for `&'a mut T`
```

Reply:
Even Stream's Next does the same https://github.com/rust-lang-nursery/futures-rs/blob/master/futures-util/src/stream/next.rs

Reply:
Ok, looking at futures-util they do the same for `Next` but from what I understand you don't need that if you can pin project.

Reply:
The bound is required because we're projecting a Pin onto a `&mut`. Without it, it would allow the body type to make some self-reference, and then still be moved after the combinator was used.

Reply:
I have no idea what you mean by pin project
So you need to elaborate if you can do it without need of `Unpin` despite requirement on Unpin in any pointer/reference

Reply:
Yeah, sorry am I at a conference so a bit slow. This is alright for ergonomic reasons since the ext trait takes a &mut self. 👍